# RAG FRAMEWORK

## a) Raw Input

Key Notes:
at 35:34
text Splitting has to be done by relevance ->embeddings_search --> topN -> chain back to LLM --> talk to AI (+relevance)

at 41:30
Okay nice, he is explanining RAG nicely :
my thoughts at that timestamp:
TECH LEVELS:
1) image + text  -> text 
    - No context: LLM only
    - eg; LLM("Hey AI, give me information on X")
2) image + text -> text  
    - Basic Contextualist: Context sample + LLM
    - eg; LLM("Hey AI, give me information on X, given (Context sample)")
3) image + text -> text 
    - Big Contextualist: Context Population + Recommendation engine + LLM
    - eg; RAG("Hey AI, give me information on X", Librarian(topN, X, Context Population))

at44:31, yes nice explanation of dimensionality reduction


at 46:55: have to classify all avialable LLMs by
matching capabilities, (based on Baijun's picture for example, CLAUDE has best reasoning --> does this imply better model embeddings?)
COST (token, contextualising cost)


at 51:38:
macthing parameters (similiarity)
cosine most popular, but could there be other techiques could have better edge at certain data?? (to investigate)



at 53:50:
VECTOR STORE
db for embeddings/vectors
to store all documents, auto gen emebddings, store auto
!!! Vector stores are optimised to do similiarity really WELL
sure anot, what optimisation, should we built everything inhouse to not be dependant. all this just funnels to get to pay for pinecone only. but yea gotta check cost-benefit here
i've already make everything up till 56mins from scratch already, we already have.
just maybe vector stores. must compare pricing and efficacy of available vector stores


at 59:04:
VECTOR STORE features
vector_store.as_retriever
connects directly to vector_store to get topN
langchain is a wrapper for LLMs. but gemini already has this feature built in.
scope: we aggregate all this LLM services. but remain portable and Microservices oriented as any one of them cld be bought over by corps (no new updates), etc

61:09 onwards is dry pinecone stuff
but yea could have features must replicate/generalise

at 1:01:33
LangChain codebase/structure highlights
prompts themselves are CLASSES, which are modularly attached to RETRIEVER_CLASS
whereas mine is functions only. Need persistence!!

at1:04:20:
langchain codebase highlights
vector_stores are wrapped in highly functionable classes
whereas mine is all objects


at 1:06:46:
Langchain UX experience and ease of "chaining":
(context + prompt + model + parser ) + prompt
so its an infinite way to arrange this.
this is basically modelling conversations with a librarian, and then to self, then writing to paper, then retalk again to librarian, then write again.

at1:09:30
!!! We need to have an AI DATA MODEL map
where we map out the AI conversations with respect to mapping to different specialist databases, human prompts, - its a conversation web modelling


at 1:12:12
wrapup:
nice video on RAG, will have to use whisper for youtube
FUTURE: can even consider other data inputs
 images (we have)
sound + video frames
UI: how? screenshot? video download? links?
AI DATA MODEL:
Process flow of conversation with Librarian and User-
prompt engineering to meet expected outputs (QA)
How to model conversations effectively (with constraints on COST/SPEED OF QUERY)
optimisation? graph theory? or even use a Graph convolutional network on top of all the web to optimise conversation path and prompt engineering?
this is athena i think and all those other company internal AI for boosting efficiency


Post video: 
IDEA:
lol if a RAG can do exactly my Youtube Keynotes taking here with analyses,
analysts will be hyperturbocharged liao
but yea depends on specialist database. so ultimately needs an SME of SME, or a specialist of the specialist database to curate this database. but yea can go even higher dimensions, reach all fields, to reach true GAI
Which is what LLM+search engines (bing copilot) aspire to do? try to be the new google across all domains, with references,
but not for just for information querying but for user experience, like Samsung AI
which is making user interface more seamless with AI augmentations.

so UI UX side, we should focus on what samsung is doing:
these are the AI funnels:

"You
galaxy ai features
Copilot
Certainly! Here are some of the features of Galaxy AI:

Circle to Search: This feature allows you to draw a circle around any object in an image, and Galaxy AI will search for relevant information related to that object.
Browsing Assist: Galaxy AI can assist with web browsing by summarizing articles, providing relevant links, and answering questions based on web content.
Transcript Assist: When you upload an audio or video file, Galaxy AI can transcribe the spoken content and provide a text-based summary.
Photo Assist: Upload an image, and Galaxy AI can describe its contents, identify objects, and provide context.
Generative Edit: Galaxy AI can help generate creative content, such as poems, stories, code, and more.
Call Assist/Live Translate: During phone calls, Galaxy AI can provide real-time translation or assist with conversation summaries.
Chat Assist: Engage in text-based conversations with Galaxy AI, which can respond contextually and provide helpful information.
Note Assist: Galaxy AI can summarize notes, generate outlines, and assist with organizing information."

then formulate all of the thinking (and whatever lacking) behind my chats above into a RAG - then we have an AI business analyst LOL




In [ ]:
keynotes_log = """
Key Notes:
at 35:34
text Splitting has to be done by relevance ->embeddings_search --> topN -> chain back to LLM --> talk to AI (+relevance)

at 41:30
Okay nice, he is explanining RAG nicely :
my thoughts at that timestamp:
TECH LEVELS:
1) image + text  -> text 
    - No context: LLM only
    - eg; LLM("Hey AI, give me information on X")
2) image + text -> text  
    - Basic Contextualist: Context sample + LLM
    - eg; LLM("Hey AI, give me information on X, given (Context sample)")
3) image + text -> text 
    - Big Contextualist: Context Population + Recommendation engine + LLM
    - eg; RAG("Hey AI, give me information on X", Librarian(topN, X, Context Population))

at44:31, yes nice explanation of dimensionality reduction


at 46:55: have to classify all avialable LLMs by
matching capabilities, (based on Baijun's picture for example, CLAUDE has best reasoning --> does this imply better model embeddings?)
COST (token, contextualising cost)


at 51:38:
macthing parameters (similiarity)
cosine most popular, but could there be other techiques could have better edge at certain data?? (to investigate)



at 53:50:
VECTOR STORE
db for embeddings/vectors
to store all documents, auto gen emebddings, store auto
!!! Vector stores are optimised to do similiarity really WELL
sure anot, what optimisation, should we built everything inhouse to not be dependant. all this just funnels to get to pay for pinecone only. but yea gotta check cost-benefit here
i've already make everything up till 56mins from scratch already, we already have.
just maybe vector stores. must compare pricing and efficacy of available vector stores


at 59:04:
VECTOR STORE features
vector_store.as_retriever
connects directly to vector_store to get topN
langchain is a wrapper for LLMs. but gemini already has this feature built in.
scope: we aggregate all this LLM services. but remain portable and Microservices oriented as any one of them cld be bought over by corps (no new updates), etc

61:09 onwards is dry pinecone stuff
but yea could have features must replicate/generalise

at 1:01:33
LangChain codebase/structure highlights
prompts themselves are CLASSES, which are modularly attached to RETRIEVER_CLASS
whereas mine is functions only. Need persistence!!

at1:04:20:
langchain codebase highlights
vector_stores are wrapped in highly functionable classes
whereas mine is all objects


at 1:06:46:
Langchain UX experience and ease of "chaining":
(context + prompt + model + parser ) + prompt
so its an infinite way to arrange this.
this is basically modelling conversations with a librarian, and then to self, then writing to paper, then retalk again to librarian, then write again.

at1:09:30
!!! We need to have an AI DATA MODEL map
where we map out the AI conversations with respect to mapping to different specialist databases, human prompts, - its a conversation web modelling


at 1:12:12
wrapup:
nice video on RAG, will have to use whisper for youtube
FUTURE: can even consider other data inputs
 images (we have)
sound + video frames
UI: how? screenshot? video download? links?
AI DATA MODEL:
Process flow of conversation with Librarian and User-
prompt engineering to meet expected outputs (QA)
How to model conversations effectively (with constraints on COST/SPEED OF QUERY)
optimisation? graph theory? or even use a Graph convolutional network on top of all the web to optimise conversation path and prompt engineering?
this is athena i think and all those other company internal AI for boosting efficiency


Post video: 
IDEA:
lol if a RAG can do exactly my Youtube Keynotes taking here with analyses,
analysts will be hyperturbocharged liao
but yea depends on specialist database. so ultimately needs an SME of SME, or a specialist of the specialist database to curate this database. but yea can go even higher dimensions, reach all fields, to reach true GAI
Which is what LLM+search engines (bing copilot) aspire to do? try to be the new google across all domains, with references,
but not for just for information querying but for user experience, like Samsung AI
which is making user interface more seamless with AI augmentations.

so UI UX side, we should focus on what samsung is doing:
these are the AI funnels:

"You
galaxy ai features
Copilot
Certainly! Here are some of the features of Galaxy AI:

Circle to Search: This feature allows you to draw a circle around any object in an image, and Galaxy AI will search for relevant information related to that object.
Browsing Assist: Galaxy AI can assist with web browsing by summarizing articles, providing relevant links, and answering questions based on web content.
Transcript Assist: When you upload an audio or video file, Galaxy AI can transcribe the spoken content and provide a text-based summary.
Photo Assist: Upload an image, and Galaxy AI can describe its contents, identify objects, and provide context.
Generative Edit: Galaxy AI can help generate creative content, such as poems, stories, code, and more.
Call Assist/Live Translate: During phone calls, Galaxy AI can provide real-time translation or assist with conversation summaries.
Chat Assist: Engage in text-based conversations with Galaxy AI, which can respond contextually and provide helpful information.
Note Assist: Galaxy AI can summarize notes, generate outlines, and assist with organizing information."

then formulate all of the thinking (and whatever lacking) behind my chats above into a RAG - then we have an AI business analyst LOL



"""

## b) Librarian

Librarian is a data-centric conversational model (RAG).

It is able to embody archetypes (specialists) and is able to converse with the user to retrieve information from these databases.

Such a data model centric design allows for the user to query information from a variety of databases, and the librarian is able to retrieve and cross-reference information from these databases.

In [7]:
# presentation 
# when client talking --> go through what they're facing 
# slides as guide/ anchor point for attention
# eg;
# highlight unstructuredness, 
# eg; go straight to the point, not loop around, blabber mode - drowning other voices with sme, - shift convo to the audience 
# listening + giving views at the same point 
# regurgation only but no effective todos, engagement- effective points
# 1) presenting ideas across 
# 2) listening to painpoints of customer 
# 3) effective address painpoints (to induce hope) 
# 4) where do we go next 

In [1]:
from RAG.librarian import Librarian 
"""
these are my inputs (showcase data set)
- layman words
- effective comms - avoid gravitation wells
"""

librarian = Librarian(librarian_LLM_model = "GEMINI")

# SELECT SPECIALIST DATABASE
librarian.select_specialist(specialist = "traveller", specialist_LLM_model = "GEMINI")

# Ask librarian to get acquinted with the specialist database
librarian.Traveller.load_data_model(reembed = False,
                                    embed_id = 0,
                                    data_model_keys = {"TEST - CLIENT":"CLIENT ID",
                                                        "TEST - CLIENT REQUEST":"CLIENT ID",
                                                        "TEST - FLIGHTS":"FLIGHT ID",
                                                        "TEST - ACCOMODATIONS":"ACCOMODATION ID",
                                                        "TEST - SERVICES":"SERVICE ID",
                                                        }
                                    )

# Data model has to be properly formulated (with their nuanced buckets) to fit their needs (quality of services, eg)
# Traveller specialist class is customised specifically for the travel industry
# - with skills that are relevant to the travel industry

loading specialist: TRAVELLER ...
TRAVELLER loaded


### i) Librarian's Traveller CAPABILITIES

### a) _Retrieval_ : client information (I)

In [8]:
prompt = "Shaik"
client_recommendations = librarian.Traveller.I_recommend_client(content = prompt,
                                                        topN=3,
                                                        # task_type = "retrieval_document",
                                                        task_type = "retrieval_query",
                                                        ) 
print(client_recommendations)

ITEM: 33, PREPARED BY: ANIES, ISSUED DATE: 2024-02-27 00:00:00, INVOICE: 2177, CLIENT ID: C139, CLIENT: DATO SRI SHAIK AQMAL BIN SHAIK ALAUDDIN, DESCRIPTION: AIR TICKET CGK-KUL 27-Feb-2024, AMOUNT (MYR): 2150.0, CN: nan, PAYMENT (RM): nan, PAYMENT DATE: nan, BAL PAYMENT: nan, COS - AIR TIX: 2150.0, COS - HOTEL: nan, COS - TRANSPORT: nan, COS-GROUND PACKAGE: nan, COS - TOUR GUIDE: nan, COS - INSURANCE: nan, COS - MISC: nan, PARTICULARS: nan, REMARKS: nan
ITEM: 35, PREPARED BY: ANIES, ISSUED DATE: 29-Feb-24, INVOICE: 2179, CLIENT ID: C139, CLIENT: DATO SRI SHAIK AQMAL BIN SHAIK ALAUDDIN, DESCRIPTION: AIR TICKET + ACCOMMODATION JKT FROM 29 FEB - 01 MAR 2024, AMOUNT (MYR): 6040.0, CN: nan, PAYMENT (RM): nan, PAYMENT DATE: nan, BAL PAYMENT: nan, COS - AIR TIX: nan, COS - HOTEL: nan, COS - TRANSPORT: nan, COS-GROUND PACKAGE: nan, COS - TOUR GUIDE: nan, COS - INSURANCE: nan, COS - MISC: nan, PARTICULARS: nan, REMARKS: nan
ITEM: 29, PREPARED BY: ANIES, ISSUED DATE: 08 Feb 24, INVOICE: 2173, CL

In [9]:
prompt = "Aqmal"
client_recommendations = librarian.Traveller.I_recommend_client_request(content = prompt,
                                                        topN=1,
                                                        # task_type = "retrieval_document",
                                                        task_type = "retrieval_query",
                                                        ) 
print(client_recommendations)

CLIENT ID: C139, DATE: 2014-03-24 00:00:00, LOCATION: kuala lumpur, PROMPT: dunno, just plan something exciting and cultural with calm vibe around an important AI event of mine on december 24. So i need to be mentally sharp before, and not shagged, DURATION: 1 month, DATE_PERIOD: 2024-12-12 00:00:00, BUDGET: 40000, PAX: 2, CLIENT QUIRKS: doesnt like humid weather and will always prefer somewhere cold, SPECIAL REQUESTS: Needs 2 bodyguards at all times


### b) _Recommend_ : travel logistics (II)

In [7]:
# Ask librarian a question regarding one of the specialist databases WITH a chatbot wrapper
# input by agent 
# FUTURE: email thread, entire convo packages - given relevant buckets, eg; dest, duration , budget

# TODO LATER: need relevance score to cut out irrelevant data
travel_proposal = "i need to go Singapore"
convo_package = librarian.Traveller.II_recommend_travel_logistics(travel_proposal = travel_proposal,
                                                                   topN=3,
                                                                   chatbot = False, 
                                                                   chatbot_model_name = "gemini-pro")

# Highlight: these outputs are from YOUR preferred buckets (your own inventory) but using gemini/openai


('flights:\n'
 'FLIGHT ID: 21, Name: Singapore Airlines, Type: Flight, From: Michigan, '
 'Singapore, To: Kuala Lumpur, Price: from $1200, Description: 15 hours '
 'Nonstop\n'
 'FLIGHT ID: 21, Name: Singapore Airlines, Type: Flight, From: Kuala Lumpur, '
 'To: Michigan, Singapore, Price: from $1200, Description: 15 hours Nonstop\n'
 'FLIGHT ID: 22, Name: Malaysia Airlines, Type: Flight, From: Michigan, '
 'Singapore, To: Singapore, Price: from $1300, Description: 15 hours Nonstop')
('accomodations:\n'
 'ACCOMODATION ID: 7, Name: Raffles hotel, Type: Hotel, Location: Singapore, '
 'Price/night: 200, Description: very close to town and access to all cultural '
 'stuff\n'
 'ACCOMODATION ID: 9, Name: Mandarin Oriental, Type: Hotel, Location: '
 'Singapore, Price/night: 2000, Description: 5 star hotel services, with 5 '
 'star breakfast and dinner but no oyster \n'
 'ACCOMODATION ID: 8, Name: Marina Bay sands, Type: Hotel, Location: '
 'Singapore, Price/night: 341, Description: Nice views o

### b) _*Generate*_ : travel proposal (itinerary) from customer request

In [5]:
# Ask librarian a question regarding one of the specialist databases
travel_proposal_request = "i need to go Singapore for a week, like in this image."
convo_package = librarian.Traveller.II_generate_travel_proposal(input_prompt = travel_proposal_request,
                                                                model_name = "gemini-pro",
                                                                image_path = "./database/travel/sometiktokss.jpg")


(' The image is of a street in Singapore, with a mosque in the background. The '
 'street is lit up by colorful lights. The image was taken at night.\n'
 '\n'
 'The location is: Haji Lane, Singapore\n'
 '\n'
 'Description: This street in Singapore is a vibrant and colorful place to '
 'visit, especially at night. The Haji Lane is a narrow street that is lined '
 'with shops, restaurants, and bars. It is a popular place to visit for both '
 'locals and tourists. The street is also known for its street art, with many '
 'of the buildings being covered in murals.\n'
 '\n'
 'You could visit the following places in Singapore:\n'
 '\n'
 'Marina Bay Sands: This is an iconic hotel and casino in Singapore. It is '
 'known for its unique architecture and its rooftop infinity pool.\n'
 'Gardens by the Bay: This is a beautiful park in Singapore that is home to '
 'many different types of plants and flowers.\n'
 'The Singapore Zoo: This zoo is home to a wide variety of animals, including '
 'pandas

In [6]:
# Ask librarian a question regarding one of the specialist databases
travel_proposal_request = str(convo_package)
convo_package = librarian.Traveller.III_generate_travel_package(travel_proposal = travel_proposal_request,
                                                                 topN = 3, 
                                                                 model_name = "gemini-pro")

# TODO: disintegrate services into 3 buckets so that fit into a day or half  day, etc
# eg; tour: 3h cultural tour at arab str, 4h foodie tour, 6h hiking tour, 3 hr shopping tour 
# - HOW TO CLASSIFY THESE BUCKETS? best way? gotta see existing travel packages (and how they are classified)
# - have to check with SME on data model - edit, modify buckets based on SmartWorld specialised needs
# prompt engineering: include a free-and easy day if theres a gap in the schedule or Reserve 1 day for RR, etc

# Customizable itinerary output - this is just a base template for them to build on 
# TODO: regenerate and recustomize BASED on revised prompt "i dont want any shopping trips, on this"

('flights:\n'
 'FLIGHT ID: 21, Name: Singapore Airlines, Type: Flight, From: Michigan, '
 'Singapore, To: Kuala Lumpur, Price: from $1200, Description: 15 hours '
 'Nonstop\n'
 'FLIGHT ID: 21, Name: Singapore Airlines, Type: Flight, From: Kuala Lumpur, '
 'To: Michigan, Singapore, Price: from $1200, Description: 15 hours Nonstop\n'
 'FLIGHT ID: 10, Name: AirAsia Berhad , Type: Flight, From: Singapore, To: '
 'Kuala Lumpur, Price: from $70 , Description: 1h 15m Nonstop')
('accomodations:\n'
 'ACCOMODATION ID: 8, Name: Marina Bay sands, Type: Hotel, Location: '
 'Singapore, Price/night: 341, Description: Nice views of the gardens by the '
 'bay and the bay\n'
 'ACCOMODATION ID: 7, Name: Raffles hotel, Type: Hotel, Location: Singapore, '
 'Price/night: 200, Description: very close to town and access to all cultural '
 'stuff\n'
 'ACCOMODATION ID: 10, Name: Mana mana suites, Type: Hotel, Location: Kuala '
 'Lumpur, Malaysia, Price/night: 70, Description: Long lifts but excellent '
 'city v

In [ ]:
# # SELECT SPECIALIST DATABASE
# librarian.select_specialist(specialist = "muslim", specialist_LLM_model = "GEMINI")
# librarian.Muslim.load_data_model(reembed = False,
#                                     data_model_keys = {"Quran":"SURAH ID",
#                                                         "Hadith":"HADITH ID",
#                                                         "Tafsir":"TAFSIR ID",
#                                                         }
#                                     )

# KnowledgeBase - Financial Data Time Series
LLM image analysis --> analyse_travel_destination --> Concierge


## a) PROMPT input

In [ ]:
date = "2024-03-04"
prompt_0 = f"""
What is the current price of Bitcoin for today ({date})? What is the latest financial news, outlook on the crypto market, 
and also contrast with the traditional stock market and money market.
"""
from settings import GEMINI_API_KEY
from llm_handler.GHandler import GHandler

g_handler = GHandler(GEMINI_API_KEY)

response = g_handler.prompt(prompt_0)
print(response.text)

## b) Image input: Multimodal LLM image analysis

In [ ]:
# image_path = "./database/quant/btcusdt_1d_IST.png"
image_path = "./database/quant/buddy_chinese.png"

prompt_1 = f"""
            you are an expert AI generated image analyst- in a AI ethics and misuse department.
            You are given images and asked to analyse them if it is AI generated or not.
            And give your reasonings.
            """
prompt_2 = f"""If the image is AI generated, provide a confidence score from 0 to 100,
            and recreate the prompt that generated the image. 
            If the image is not AI generated, then craft a single paragraph, descriptive prompt that could result in a similar image.
"""
# prompt_2v = f"""If the image is AI generated, provide a confidence score from 0 to 100,
# "The expert AI generated image analyst response to the prior prompt ({prompt_1}) is: 
# <trader's RESPONSE>".
# "My own thoughts on his response is: 
# <vibe_checker's RESPONSE>".
# "Confidence: <YOUR_SCORE>".
# """
from IPython.display import Image, display
display(Image(filename=image_path))
from pprint import pprint
from settings import GEMINI_API_KEY
from llm_handler.GHandler import GHandler

g_handler = GHandler(GEMINI_API_KEY)

g_response = g_handler.prompt_image(model_name = "gemini-pro-vision",
                                  image_path = image_path,
                                  prompt_1 = prompt_1,
                                  prompt_2 = prompt_2,
                                  generation_config = {
                                                        "temperature": 0.9,
                                                        "top_p": 0.95,
                                                        "top_k": 40,
                                                        "max_output_tokens": 1024,
                                                        },
                                block_threshold = "BLOCK_NONE"
                                        )
# print this in a nicer format and not stretched beyond the screen
pprint(g_response.text)

In [ ]:

Generate an image with the following description: 
 "The photograph is of a gray tabby cat being held by a person. The cat's eyes "
 'are partially closed and it looks relaxed. The photograph is taken from a '
 'close-up angle and the background is out of focus.\n'
 '\n'
 'To take a similar photograph, you would need a camera and a cat. You would '
 'need to get close to the cat and take the photograph when it is relaxed. You '
 'could also use a telephoto lens to get a closer shot.')

COMMENTS

The code snippet you provided attempts to utilize the Gemini-Pro-Vision model for image analysis in the context of a financial time series chart. However, there are several limitations and concerns to consider:

**Limitations:**

1. **Model capabilities:** While Gemini-Pro-Vision is trained on a massive dataset of text and code, it isn't specifically designed for analyzing financial charts. Its ability to interpret the image and provide the specific descriptions, diagnoses, predictions, and prescriptions you require might be limited.

2. **Data complexity:** Financial charts contain complex information encoded in lines, colors, and patterns. Capturing this information and translating it into accurate financial insights requires specialized training data and models.

**Concerns:**

1. **Over-reliance on LLM:** The code fully relies on the LLM's interpretation of the image, leaving no room for human expertise or domain knowledge. This can lead to inaccurate or misleading financial insights.

2. **Ethical considerations:** Using an LLM for financial predictions without proper validation and transparency can be risky. It's essential to be clear about the limitations of the system and emphasize that its output should not be solely relied upon for investment decisions.

3. **Potential misuse:** This approach might encourage users to interpret the LLM's output as financial advice, even though it might lack the necessary accuracy and context.

**Alternative Approach:**

Instead of relying solely on an LLM, consider a more **hybrid approach**:

1. **Human Analysis:** Involve a human financial analyst to analyze the chart and provide their expertise. This can help ensure the accuracy and reliability of the initial assessment.

2. **LLM Support:** Use the LLM to **complement the human analysis** by:
    * Summarizing relevant financial news or historical data related to the instrument.
    * Providing alternative perspectives or highlighting potential risks.
    * Generating different creative writing formats (e.g., a news article about the market situation) based on the analyst's insights.

This approach leverages the strengths of both humans and LLMs, potentially leading to more robust and reliable insights while mitigating ethical concerns.

It's crucial to remember that LLMs are still under development, and their capabilities in specialized domains like finance are evolving. While they can be valuable tools, exercising caution and combining them with human expertise remains essential, especially when dealing with financial decision-making.

## c) Inventory Recommendation Engine


i) Embed database

ii) embed user_travel_LLM_response above

iii) AI recommendation engine


### i) Embed -> Recommend Inventory

In [ ]:
db_path = "./database/travel/Data Model for Travel.xlsx"
sheet_name = "Day Trip"

from settings import GEMINI_API_KEY
from llm_handler import GHandler
import importlib
importlib.reload(GHandler)
from llm_handler.GHandler import GHandler

import pandas as pd

day_trips_df = pd.read_excel(db_path, sheet_name=sheet_name)
# need to embed the day_trips_df
day_trips_df["Text"] = day_trips_df.apply(lambda row: f"Activity: {row['Activity']}, Location: {row['Location']}, Category: {row['Category']} Price: {row['Price']}", axis=1)


g_handler = GHandler(GEMINI_API_KEY)
df_embedded = g_handler.embed_df(day_trips_df,
                                 title = "Activity", 
                                 text = "Text",
                                 model="models/embedding-001")
daytrip_recommendation = g_handler.find_best_passage(g_response.text, df_embedded)
print(daytrip_recommendation)

## d) Cascade Recommendation Engine

Given the user's primary match, the recommendation engine will look up secondary and tertiary recommendations


In [ ]:
# Get embeddings for the hotels and flights
flights_hotels_df = pd.read_excel(r'./database/travel/Data Model for Travel.xlsx', sheet_name="Hotels")
# drop where type == Nan
flights_hotels_df = flights_hotels_df.dropna(subset=["Type"])
flights_hotels_df["Text"] = flights_hotels_df.apply(lambda row: f"Name: {row['Name']}, Type: {row['Type']}, Location: {row['Location']}, Price: {row['Price']}, Description: {row['Description']}", axis=1)


g_handler = GHandler(GEMINI_API_KEY)
flights_hotels_df = g_handler.embed_df(flights_hotels_df,
                                 title = "Name", 
                                 text = "Text",
                                 model="models/embedding-001")


In [ ]:
Convo_2 = f"""I have an inventory for a {daytrip_recommendation}. I can recommend you a more luxurious based on the location of the activity."""
hotel_recommendation = g_handler.find_best_passage(Convo_2, flights_hotels_df)
print(hotel_recommendation)

In [ ]:
Convo_3 = f"""I have an inventory for a {daytrip_recommendation}. I can recommend you the longest flight based on the location of the activity."""
flight_recommendation = g_handler.find_best_passage(Convo_3, flights_hotels_df)
print(flight_recommendation)

In [ ]:
flights_hotels_df

In [ ]:
# Wrapper that inserts
# daytrip + hotel + flight into a 3 day 2 night itenerary
# so first day fill with casual eating and sightseeing, then 2nd day is the daytrip focus, and 3rd day is the return flight
# LLM has to be smart enough to infer duration of daytrip (hiking mt rinjani is 2 days, so it should be 2 days 1 night)

# Eg: himalayas
# has to account for weather, seasons, and other factors that might affect the trip duration. Or postpone trip under more optimal factors
# TESTCASE: NO inventory for customer request
# TESTCASE: transport types (ferries, etc)

# look at all factor buckets that involves inventory 
# Buckets to quantify the inventory
# main factor is affiliation, services (driver, villa-cook, guide, translator), then location, then price, then type, then rating, then availability, then duration, then weather, then season, then other factors)
# if not affiliated then shouldnt be recommended 




In [ ]:
# top n recommendations (pro consumer vs pro business)

# need to see if they have existing monetisation strategy for their travel packages
# recommmendations should be tailored to their monetisation strategy
# (eg: if they have a partnership with a hotel, then recommend that hotel)


# main objective: recommendation engine 
# warning: no buckets for existing monetisation strategy

In [ ]:
daytrip_recommendation

In [ ]:
flight_recommendation

In [ ]:
flight_recommendation

In [ ]:
UX_prompt = f"""
Generate a travel package for the given trip_recommendation:
{daytrip_recommendation}

This trip_recommendation comes with the following hotel recommendation:
{hotel_recommendation}

This trip_recommendation comes with the following flight recommendation:
{flight_recommendation}

The package should include the following sections:

"Summary"
introductory and summary of the trip in one paragraph.
The summary should describe in vivid detail, the main attractions, activities, and experiences that the travelers can enjoy in the trip_recommendation. 

"Journey Highlights"
A list of the main features and most exciting aspects of the package.
the highlights must end off with a bold line: "Your journey takes you to: x - y - z"

"Itinerary & Map" 
This section is an itenerary list that shows the day-by-day plan of the trip, that is also accompanied by a map.

The itinerary should include the name, location, and description of each place or activity that the travelers will visit or do each day. 
The itinerary should also indicate the approximate duration and transportation mode for each item.

A highlights and inclusions section that lists the main features and benefits of the package. 
The section should mention what is included in the price, such as flights, accommodation, meals, guides, entrance fees, etc. 
The section should also mention any special offers or discounts that are available for the package.
A dates and pricing section that shows the available dates and prices for the package. 
The section should indicate the departure and return dates, the number of travelers, the total cost, and the payment options for the package. 
The section should also provide a link or contact information for booking or inquiring about the package.

All information derived here should be based on the recommendations from the previous steps and MUST not be fabricated.

"""

In [ ]:
from settings import GEMINI_API_KEY
from llm_handler.GHandler import GHandler

g_handler = GHandler(GEMINI_API_KEY)

response_UX = g_handler.prompt(UX_prompt)
print(response_UX.text)

In [ ]:
# Format mimic from image of travel website package page

In [ ]:
image_path = "./database/travel/lux_example.jpg"
prompt_1 = f"""Reproduce the format of this travel package given a different destination: 
Generate a travel package for the given trip_recommendation:
{daytrip_recommendation}

This trip_recommendation comes with the following hotel recommendation:
{hotel_recommendation}

This trip_recommendation comes with the following flight recommendation:
{flight_recommendation}

"""

from IPython.display import Image, display
# display(Image(filename=image_path))

from settings import GEMINI_API_KEY
from llm_handler.GHandler import GHandler

g_handler = GHandler(GEMINI_API_KEY)

g_response = g_handler.prompt_image(model_name = "gemini-pro-vision",
                                  image_path = image_path,
                                  prompt_1 = prompt_1,
                                  prompt_2 = None)
print(g_response.text)

## e) Full Stack

In [ ]:
image_path = "./database/travel/me-at-kelingking-beach-nusa-penida-bali-inonesia-laugh-traveleat.jpg"
prompt_1 = "Tell me the location where this photo is taken from?"
prompt_2 = "Based on the response, recommend a full day trip travel itinerary"
from IPython.display import Image, display
print("(a) Image Analyses")
display(Image(filename=image_path))


import pandas as pd
from settings import GEMINI_API_KEY
from llm_handler.GHandler import GHandler

g_handler = GHandler(GEMINI_API_KEY)

g_response = g_handler.prompt_image(model_name = "gemini-pro-vision",
                                  image_path = image_path,
                                  prompt_1 = prompt_1,
                                  prompt_2 = prompt_2)
# print(g_response.text)
print("(b) Image Analyses Complete")

db_path = "./database/travel/Data Model for Travel.xlsx"
sheet_name = "Day Trip"

def get_day_trip_recommendation(g_response, db_path, sheet_name):
    day_trips_df = pd.read_excel(db_path, sheet_name=sheet_name)
    # need to embed the day_trips_df
    day_trips_df["Text"] = day_trips_df.apply(lambda row: f"Activity: {row['Activity']}, Location: {row['Location']}, Category: {row['Category']} Price: {row['Price']}", axis=1)


    g_handler = GHandler(GEMINI_API_KEY)
    df_embedded = g_handler.embed_df(day_trips_df,
                                    title = "Activity", 
                                    text = "Text",
                                    model="models/embedding-001")
    daytrip_recommendation = g_handler.find_best_passage(g_response.text, df_embedded)
    return daytrip_recommendation

In [ ]:
db_path = "./database/travel/Data Model for Travel.xlsx"
sheet_name = "Day Trip"

from settings import GEMINI_API_KEY
from llm_handler import GHandler
import importlib
importlib.reload(GHandler)
from llm_handler.GHandler import GHandler

import pandas as pd

day_trips_df = pd.read_excel(db_path, sheet_name=sheet_name)
# need to embed the day_trips_df
day_trips_df["Text"] = day_trips_df.apply(lambda row: f"Activity: {row['Activity']}, Location: {row['Location']}, Category: {row['Category']} Price: {row['Price']}", axis=1)


g_handler = GHandler(GEMINI_API_KEY)
df_embedded = g_handler.embed_df(day_trips_df,
                                 title = "Activity", 
                                 text = "Text",
                                 model="models/embedding-001")
daytrip_recommendation = g_handler.find_best_passage(g_response.text, df_embedded)
print(daytrip_recommendation)

# Talk to your Fitra AI 

So imagine you are walking back home, from work, on your phone, 
Then you see a website with some inspiring duas, and you want to analyse and 
understand the meaning of the duas, and you want to know the meaning of the duas,
and find contextually similiar duas or information. 
eg;
- "Rabbana atina fid-dunya hasanatan wa fil 'akhirati hasanatan waqina 'adhaban-nar"
- "Our Lord, give us in this world [that which is] good and in the Hereafter [that which is] good and protect us from the punishment of the Fire."
- then you want to find similar duas, or background information about this dua like its origins, hadith chains, and all that 





TECH LEVELS:
1) image + text  -> text 
    - No context: LLM only
    - eg; LLM("Hey AI, give me information on X")
2) image + text -> text  
    - Basic Contextualist: Context sample + LLM
    - eg; LLM("Hey AI, give me information on X, given (Context sample)")
3) image + text -> text 
    - Big Contextualist: Context Population + Recommendation engine + LLM
    - eg; RAG("Hey AI, give me information on X", Librarian(topN, X, Context Population))

In [ ]:
from specialists.muslim import Muslim 
muslim = Muslim()
muslim.load_data_model(reembed = True)

In [ ]:
text = muslim.tables["Sunan al Tirmidhi"].iloc[1,1]
text

In [ ]:
test= muslim.model_specialist.embed_text(title= "hadith", text= text, model="models/embedding-001")

In [ ]:
import pandas as pd

from llm_handler.GHandler import GHandler
from settings import GEMINI_API_KEY

g_handler = GHandler(GEMINI_API_KEY)

db_path = "./database/muslim/Sunan al Tirmidhi.csv"
df = pd.read_csv(db_path)
df.iloc[0]

#  translate df.iloc[0] from arabic to english
response = g_handler.prompt(f"Please translate the following from arabic to english: It is from a hadith: {df.iloc[0,0]}")

In [ ]:
from pprint import pprint
pprint(response.text)

In [ ]:
df

## Tech tier I: Basic Contextualist

In [ ]:
# image_path = "./database/quant/btcusdt_1d_IST.png"
image_path = "./database/Fitra AI/dua_1.png"

prompt_1 = f"""
            You are a curator and specialist of all knowledge regarding Islam- 
            in a AI ethics and misuse department.
            You are given images and asked to analyse them for their Islamic content and also level of AI influence.
            So look firstly, for Islamic data authenticity then after that, if there are any AI generated content,
            especially in meanings or translations, give your reasonings.
            """
prompt_2 = f"""
            If the image is not AI generated and is authentic Islamic content, 
            then suggest a list of relevant topics to the Islamic content in this image,
            such as its origins, hadith sources, Quran sources, and other relevant Islamic topics.
"""
# prompt_2v = f"""If the image is AI generated, provide a confidence score from 0 to 100,
# "The expert AI generated image analyst response to the prior prompt ({prompt_1}) is: 
# <trader's RESPONSE>".
# "My own thoughts on his response is: 
# <vibe_checker's RESPONSE>".
# "Confidence: <YOUR_SCORE>".
# """
from IPython.display import Image, display
display(Image(filename=image_path))
from pprint import pprint
from settings import GEMINI_API_KEY
from llm_handler.GHandler import GHandler

g_handler = GHandler(GEMINI_API_KEY,
                     generation_config = {"temperature": 0.9,
                                      "top_p": 0.95,
                                      "top_k": 40,
                                      "max_output_tokens": 1024,
                                      },
                     block_threshold="BLOCK_NONE",
                    )

g_response = g_handler.prompt_image(model_name = "gemini-pro-vision",
                                  image_path = image_path,
                                  prompt_1 = prompt_1,
                                  prompt_2 = prompt_2,)
# print this in a nicer format and not stretched beyond the screen
pprint(g_response.text)

## Tech tier II: Small Contextualist


i) Embed database

ii) embed user_travel_LLM_response above

iii) AI recommendation engine


### i) Embed -> Recommend Inventory

In [ ]:
db_path = "./database/travel/SWTT_ Master Database.xlsx"
sheet_name = "CUSTOMERS"

from settings import GEMINI_API_KEY
from llm_handler import GHandler
import importlib
importlib.reload(GHandler)
from llm_handler.GHandler import GHandler

import pandas as pd
tabs = pd.ExcelFile(db_path).sheet_names 
tabs = [sheet for sheet in tabs if "TEST" in sheet]
print(tabs)
# df = pd.read_excel(db_path, sheet_name=sheet_name)
# # need to embed the day_trips_df
# day_trips_df["Text"] = day_trips_df.apply(lambda row: f"Activity: {row['Activity']}, Location: {row['Location']}, Category: {row['Category']} Price: {row['Price']}", axis=1)


# g_handler = GHandler(GEMINI_API_KEY)
# df_embedded = g_handler.embed_df(day_trips_df,
#                                  title = "Activity", 
#                                  text = "Text",
#                                  model="models/embedding-001")
# daytrip_recommendation = g_handler.find_best_passage(g_response.text, df_embedded)
# print(daytrip_recommendation)

## d) Cascade Recommendation Engine

Given the user's primary match, the recommendation engine will look up secondary and tertiary recommendations


In [ ]:
# Get embeddings for the hotels and flights
flights_hotels_df = pd.read_excel(r'./database/travel/Data Model for Travel.xlsx', sheet_name="Hotels")
# drop where type == Nan
flights_hotels_df = flights_hotels_df.dropna(subset=["Type"])
flights_hotels_df["Text"] = flights_hotels_df.apply(lambda row: f"Name: {row['Name']}, Type: {row['Type']}, Location: {row['Location']}, Price: {row['Price']}, Description: {row['Description']}", axis=1)


g_handler = GHandler(GEMINI_API_KEY)
flights_hotels_df = g_handler.embed_df(flights_hotels_df,
                                 title = "Name", 
                                 text = "Text",
                                 model="models/embedding-001")


In [ ]:
Convo_2 = f"""I have an inventory for a {daytrip_recommendation}. I can recommend you a more luxurious based on the location of the activity."""
hotel_recommendation = g_handler.find_best_passage(Convo_2, flights_hotels_df)
print(hotel_recommendation)

In [ ]:
Convo_3 = f"""I have an inventory for a {daytrip_recommendation}. I can recommend you the longest flight based on the location of the activity."""
flight_recommendation = g_handler.find_best_passage(Convo_3, flights_hotels_df)
print(flight_recommendation)

In [ ]:
flights_hotels_df

In [ ]:
# Wrapper that inserts
# daytrip + hotel + flight into a 3 day 2 night itenerary
# so first day fill with casual eating and sightseeing, then 2nd day is the daytrip focus, and 3rd day is the return flight
# LLM has to be smart enough to infer duration of daytrip (hiking mt rinjani is 2 days, so it should be 2 days 1 night)

# Eg: himalayas
# has to account for weather, seasons, and other factors that might affect the trip duration. Or postpone trip under more optimal factors
# TESTCASE: NO inventory for customer request
# TESTCASE: transport types (ferries, etc)

# look at all factor buckets that involves inventory 
# Buckets to quantify the inventory
# main factor is affiliation, services (driver, villa-cook, guide, translator), then location, then price, then type, then rating, then availability, then duration, then weather, then season, then other factors)
# if not affiliated then shouldnt be recommended 




In [ ]:
# top n recommendations (pro consumer vs pro business)

# need to see if they have existing monetisation strategy for their travel packages
# recommmendations should be tailored to their monetisation strategy
# (eg: if they have a partnership with a hotel, then recommend that hotel)


# main objective: recommendation engine 
# warning: no buckets for existing monetisation strategy

In [ ]:
daytrip_recommendation

In [ ]:
flight_recommendation

In [ ]:
flight_recommendation

In [ ]:
UX_prompt = f"""
Generate a travel package for the given trip_recommendation:
{daytrip_recommendation}

This trip_recommendation comes with the following hotel recommendation:
{hotel_recommendation}

This trip_recommendation comes with the following flight recommendation:
{flight_recommendation}

The package should include the following sections:

"Summary"
introductory and summary of the trip in one paragraph.
The summary should describe in vivid detail, the main attractions, activities, and experiences that the travelers can enjoy in the trip_recommendation. 

"Journey Highlights"
A list of the main features and most exciting aspects of the package.
the highlights must end off with a bold line: "Your journey takes you to: x - y - z"

"Itinerary & Map" 
This section is an itenerary list that shows the day-by-day plan of the trip, that is also accompanied by a map.

The itinerary should include the name, location, and description of each place or activity that the travelers will visit or do each day. 
The itinerary should also indicate the approximate duration and transportation mode for each item.

A highlights and inclusions section that lists the main features and benefits of the package. 
The section should mention what is included in the price, such as flights, accommodation, meals, guides, entrance fees, etc. 
The section should also mention any special offers or discounts that are available for the package.
A dates and pricing section that shows the available dates and prices for the package. 
The section should indicate the departure and return dates, the number of travelers, the total cost, and the payment options for the package. 
The section should also provide a link or contact information for booking or inquiring about the package.

All information derived here should be based on the recommendations from the previous steps and MUST not be fabricated.

"""

In [ ]:
from settings import GEMINI_API_KEY
from llm_handler.GHandler import GHandler

g_handler = GHandler(GEMINI_API_KEY)

response_UX = g_handler.prompt(UX_prompt)
print(response_UX.text)

In [ ]:
# Format mimic from image of travel website package page

In [ ]:
image_path = "./database/travel/lux_example.jpg"
prompt_1 = f"""Reproduce the format of this travel package given a different destination: 
Generate a travel package for the given trip_recommendation:
{daytrip_recommendation}

This trip_recommendation comes with the following hotel recommendation:
{hotel_recommendation}

This trip_recommendation comes with the following flight recommendation:
{flight_recommendation}

"""

from IPython.display import Image, display
# display(Image(filename=image_path))

from settings import GEMINI_API_KEY
from llm_handler.GHandler import GHandler

g_handler = GHandler(GEMINI_API_KEY)

g_response = g_handler.prompt_image(model_name = "gemini-pro-vision",
                                  image_path = image_path,
                                  prompt_1 = prompt_1,
                                  prompt_2 = None)
print(g_response.text)

## e) Full Stack

# Field populator

In [ ]:
columns = """

																																																																										ACCOMMODATION ARRANGMENTS																	NUMBER OF PAX								TRAVEL INSURANCE									VISA ARRANGEMENTS									BESPOKE VIP SERVICES							ENTIRE TRIP GRAND TOTAL (MYR)
ITEM	CLIENTELE													GENERAL BIODATA									CLIENT DIRECT BOOKING							PIC/BOOKER															TRIP							PURCHASE TYPE							FLIGHT ARRANGEMENTS															B2B PLATFORM	DESTINATION	GROUP	PROPERTY TYPE	HOTEL BRAND	STAR RATING	TRAVEL DATE	PERIOD OF STAY (NUMBER OF NIGHTS)	ROOM CATEGORY	ROOM CONFIGURATION (BED-TYPE)	ROOM UNITS	ROOM OCCUPANCY (NUMBER OF GUESTS)	ROOM RATE WITH BREAKFAST/NIGHT (MYR)		ROOM-ONLY RATE /NIGHT (MYR)		GRAND TOTAL	"ADULT 
(12 YEARS OLD AND ABOVE)"	"CHILD
(INFANT - 11 YEARS OLD)"		DOMESTIC HELPER		STAFF	BODYGUARD	POLICE ESCORT	INSURANCE COMPANY	INSURANCE TYPE	INSURANCE PLAN	TRIP PLAN	POLICY NUMBER	PERIOD OF COVERAGE	COST	SELLING	COUNTRIES WE COVER	VISA TYPE	VISA CATEGORY	APPLICANT NAME	REFERENCE NUMBER	PERIOD OF STAY	INSURED BY	COST PER PERSON	MARK UP RATE (MYR)	SELLING RATE PER PERSON	MEET & GREET	AIRPORT DUTY	BUGGY ASSISTANCE	LUGGAGE WRAP SERVICE	SPECIAL CARE ASSISTANCE	VIP LOUNGE	SMART WORLD ALLOWANCE TO STAFF	
	CLIENT ID	ORGANIZATION	SEGMENT	REMARK	PREFIX	TITLE	FULL NAME	DESIGNATION | OCCUPATION	RELATIONSHIP	CLIENT ID	COMPANY	BUSINESS ADDRESS	REGION	BIRTH DATE	PASSPORT NUMBER	IDENTIFICATION NUMBER	MARITAL STATUS	GENDER	AGE	RACE	NATIONALITY 	RELIGION	CLIENT TYPE	"CLIENT SINCE
(YEAR)"	EMAIL ADDRESS	COUNTRY CODE	PHONE	COUNTRY CODE	MOBILE	PREFIX	TITLE	BOOKER NAME	DESIGNATION/OCCUPATION	RELATIONSHIP	CLIENT ID	DEPARTMENT	COMPANY	BUSINESS ADDRESS	REGION	EMAIL ADDRESS	COUNTRY CODE	PHONE	COUNTRY CODE	MOBILE	CLIENT SPENDING POWER	PURPOSE OF TRAVEL	TRAVEL TYPE	OCCASION	FAMILY COMPOSITION	"DATE OF TRAVEL
(DD.MM.YY - DD.MM.YY)"	LENGTH OF TRAVEL (DAY)	FLIGHT	ACCOMMODATION	TRANSPORTATION	VIP SERVICE	VISA	TRAVEL INSURANCE	TOUR PACKAGE	TRIP NAME	AIRLINE NAME	CATEGORY	DESTINATION	DOMESTIC FLIGHT	INTERNATIONAL FLIGHT	ONE WAY TRIP	RETURN TICKET	DEPARTURE DATE	DEPARTURE TIME	ARRIVAL DATE	ARRIVAL TIME	COST	MARK UP RATE	SELLING													COST ROOM RATE PER NIGHT WITH BREAKFAST (MYR)	SELLING ROOM RATE PER NIGHT WITH BREAKFAST (MYR)	COST ROOM-ONLY RATE PER NIGHT (MYR)	SELLING ROOM -ONLY RATE PER NIGHT (MYR)					NATIONALITY	PAX												CLUSTER | DESTINATION																	
"""

In [ ]:
image_path = "./database/travel/travel_itinerary.jpg"
prompt_1 = "From this picture of a travel itinerary"
prompt_2 = "Based on the response, recommend a full day trip travel itinerary"
from IPython.display import Image, display
display(Image(filename=image_path))

from settings import GEMINI_API_KEY
from llm_handler.GHandler import GHandler

g_handler = GHandler(GEMINI_API_KEY)

g_response = g_handler.prompt_image(model_name = "gemini-pro-vision",
                                  image_path = image_path,
                                  prompt_1 = prompt_1,
                                  prompt_2 = prompt_2)
print(g_response.text)


Process Flow
0) customer request 
-  

1) Flight booking
- SABRE 
- This must be determined first, 


2) Hotel: B2B platforms: 
- WithinEarth
- TBO
- RAtehawk 
- booking.com 


tour package
- DMC partner lag (24 - 48hours) 

DMC partners need to be filtered, 

proposal:  0.5 days, finalis, another half of the day - hanis will compile
- check flight first, ticketing then hotel. 
STANDARD: 24-48 hours within proposal to get invoice. 

proposals usually take 24 hours, 

invoice, confirmed, wit client then draft of the invoice, --> to finance, and client. 
- depends on payment terms, if they have add on services, then invoice will be sent after trip. 
- mostly high profile boss, but no penalty. penalty 1.2% ---> 15% pay upfreont, another 10% credit term, 5% deposit, 75% the rest after 



mostly the same for umrah. 